compiler/src/iree/compiler/Dialect/Flow/Transforms/DumpDispatchGraph.cpp

```diff
@@ -0,0 +1,422 @@
+        }
+      } else {
+        operand.printAsOperand(os, state);
+      }
+
+      if (i != e - 1) {
+        os << ", ";
+      }
+    }
+  }
+
+  /// Generate a label for an operation.
+  std::string getLabel(Operation *op) {
+    return strFromOs([&](raw_ostream &os) {
+      if (op->getNumRegions() == 0) {
+        auto funcOp = op->getParentOfType<func::FuncOp>();
+        AsmState state(funcOp);
+
+        for (auto result : op->getResults()) {
+          result.printAsOperand(os, state);
+        }
+        os << " = " << op->getName();
+
+        if (auto dispatch = dyn_cast<DispatchOp>(op)) {
+          // print workgroup count
+          os << "[";
+          printOperands(os, dispatch.workgroup_count(), state);
+          os << "]\n";
+
+          // print entry function name
+          os << dispatch.entry_point();
+
+          // print entry function args
+          os << "(";
+          printOperands(os, dispatch.operands(), state);
+          os << ")\n";
+
+        } else {
+          os << "\n";
+        }
+      } else {
+        os << op->getName() << "\n";
+      }
+
+      if (printResultTypes) {
+        std::string buf;
+        llvm::raw_string_ostream ss(buf);
+        interleave(op->getResultTypes(), ss, "\n");
+        os << ss.str();
+      }
+    });
+  }
+
+  /// Generate a label for a block argument.
+  std::string getLabel(BlockArgument arg) {
+    return "arg" + std::to_string(arg.getArgNumber());
+  }
+
+  /// Process a block. Emit a cluster and one node per block argument and
+  /// operation inside the cluster.
+  void processBlock(Block &block) {
+    emitClusterStmt([&]() {
+      for (BlockArgument &blockArg : block.getArguments())
+        valueToNode[blockArg] = emitNodeStmt(getLabel(blockArg));
+
+      // Emit a node for each operation.
+      Optional<Node> prevNode;
+      for (Operation &op : block) {
+        Node nextNode = processOperation(&op);
+        if (printControlFlowEdges && prevNode)
+          emitEdgeStmt(*prevNode, nextNode, /*label=*/"",
+                       kLineStyleControlFlow);
+        prevNode = nextNode;
+      }
+    });
+  }
+
+  bool isScalarConstantOp(Operation *op) {
+    if (auto constOp = dyn_cast<mlir::arith::ConstantOp>(op))
+      if (constOp.getResult().getType().isIntOrIndexOrFloat()) return true;
+
+    return false;
+  }
+
+  /// Process an operation. If the operation has regions, emit a cluster.
+  /// Otherwise, emit a node.
+  Node processOperation(Operation *op) {
+    Node node;
+
+    if (isScalarConstantOp(op)) {
+      // don't handle scalar constant because it creates too many edges
+      // from a single constant.
+      return node;
+    }
+
+    if (op->getNumRegions() == 1) {
+      // do not generate a cluster when there is one region.
+      processRegion(op->getRegion(0));
+    } else if (op->getNumRegions() > 0) {
+      // Emit cluster for op with regions.
+      node = emitClusterStmt(
```

Comment:
Whats the level at which you are targeting the rendering. Should this work for "any" part of the IR? If this is just after dispatches, then I dont think we need cluster stmts. Its probably good to hide that recursive information when using something like graphviz. If there were a more interactive rendered (like tensorboard that could expand when clicked on) then region operations might be useful.

Reply:
Yes, I already hid multiple nested cluster boxes. I specialized the pass for after outlining dispatch regions.

Reply:
A web-based solution would be great to show much more information like the full dispatch region body code, clicking & following operands to its definition site, and so on.